Feature/added dataset docs (#518)
CitrineInformatics/citrine-python: docs/source/getting_started/datasets.rst

Datasets can only be deleted if they are empty.
If you need to delete a non-empty dataset, see this discussion on :ref:`deleting data objects <deleting_data_objects_label>`.

Dataset Creation and Deletion
-----------------------------

> A dataset can be created (or deleted) from its project using the register and delete methods. Note that the creation of a dataset requires the creation of a `Dataset` object.

Reviewer: ```suggestion
A dataset can be created (or deleted) from its project using the register and delete methods. Note that the creation of a dataset requires the creation of a ``Dataset`` object.
```

> A dataset can be created (or deleted) using the project from which it should or currently originates using the :func:`~citrine.resources.dataset.DatasetCollection.register` and :func:`~citrine.resources.project.ProjectCollection.delete` methods. Note that the creation of a dataset requires the creation of a `Dataset` object.

Reviewer: ```suggestion
A dataset can be created (or deleted) using the project from which it should or currently originates using the :func:`~citrine.resources.dataset.DatasetCollection.register` and :func:`~citrine.resources.dataset.DatasetCollection.delete` methods.
```
Based on the code example below it looks like the method is `project.datasets.delete`, not `citrine.projects.delete`.

Reviewer: What does "from which it should or currently originates" mean?

Reviewer: I suggested deleting the second sentence because the need to instantiate the object is, I think, made clear both in the next sentence ("To create a dataset...") and in the code examples.

Author: I don't think this is true. In fact, when I use the proposed path to the method, the link won't even build The delete method does appear to live in the project. https://github.com/CitrineInformatics/citrine-python/blob/49d7c8241d2c4a81438139e1d2483d8d99cc3b86/src/citrine/resources/project.py#L524

The delete method available within the dataset module is only for data objects     https://github.com/CitrineInformatics/citrine-python/blob/49d7c8241d2c4a81438139e1d2483d8d99cc3b86/src/citrine/resources/dataset.py#L281

Reviewer: There are a few delete methods here to consider.
On the [dataset resource](https://github.com/CitrineInformatics/citrine-python/blob/49d7c8241d2c4a81438139e1d2483d8d99cc3b86/src/citrine/resources/dataset.py#L281), `dataset.delete()`, this deletes a data object, and as you pointed out, that's not what we want to use here.
On the [project collection](https://github.com/CitrineInformatics/citrine-python/blob/49d7c8241d2c4a81438139e1d2483d8d99cc3b86/src/citrine/resources/project.py#L524), `citrine.projects.delete()`. That takes a project id and deletes a project.
On the [dataset collection](https://github.com/CitrineInformatics/citrine-python/blob/49d7c8241d2c4a81438139e1d2483d8d99cc3b86/src/citrine/_rest/collection.py#L113), `project.datasets.delete()`. You're right that `delete` is not explicitly defined for `DatasetCollection`, so it uses the generic `Collection` delete method. This takes a dataset id and deletes the dataset.

Reviewer: So, the code example below is correct (at least that's how it should work; I haven't tried deleting a dataset myself). You're right that the link I proposed above will not work because `delete` is not overwritten in that `DatasetCollection`. We could either link to the generic method in `Collection`, or we could forego links altogether. These specific implementations of `delete` don't involve novel signatures, nor do they have information in their docstrings that is surprising or necessary for the reader of this document to see. The code below demonstrates how to use them. So I suggest the text be simply "using the ``register`` and ``delete`` methods."

Author: Thanks! That's super useful information!

Author: I made the suggested changes in my latest commit :)

Creating a Dataset
^^^^^^^^^^^^^^^^^^

> To create a dataset, the `Dataset` object must first be instantiated, and then passed to the registration endpoint.

Reviewer: ```suggestion
To create a dataset, the ``Dataset`` object must first be instantiated, and then passed to the registration endpoint.
```

Example
^^^^^^^

> Assume you have a "band gaps project" with known ID, `band_gaps_project_id`, and are trying to make a dataset within that project.

Reviewer: ```suggestion
Assume you have a "band gaps project" with known ID, ``band_gaps_project_id``, and are trying to make a dataset within that project.
```

.. code-block:: python

    from citrine.resources.dataset import Dataset

    # get the project the dataset will belong to
    band_gaps_project = citrine.projects.get(band_gaps_project_id)

    # create the Dataset object
    Strehlow_Cook_description = "Band gaps for elemental and binary semiconductors with phase and temperature of measurement. DOI 10.1063/1.3253115"
    Strehlow_Cook_dataset = Dataset(name="Strehlow and Cook", summary="Strehlow and Cook band gaps", description=Strehlow_Cook_description)

    # pass the Dataset object to the registration endpoint
    Strehlow_Cook_dataset = band_gaps_project.datasets.register(Strehlow_Cook_dataset)

Deleting a Dataset
^^^^^^^^^^^^^^^^^^

A dataset can be deleted using the project from which it originates. Note that the dataset must be empty before deleting it. 

Example
^^^^^^^

> Assume you have a "band gaps project" with known ID, `band_gaps_project_id`, and are trying to delete a dataset within that project with known ID, `strehlow_cook_dataset_id`.

Reviewer: ```suggestion
Assume you have a "band gaps project" with known ID, ``band_gaps_project_id``, and are trying to delete a dataset within that project with known ID, ``strehlow_cook_dataset_id``.
```

    # access so that it can only be retrieved and from the band_gaps_project.
    band_gaps_project.make_private(strehlow_cook_dataset)

Sharing With a Specific Project
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> A dataset can be shared with another project using the project from which is originates using the :func:`~citrine.resources.project.Project.share` method on the project.

Reviewer: ```suggestion
A dataset can be shared with another project using the :func:`~citrine.resources.project.Project.share` method on the original project.
```

Example
^^^^^^^

> Assume you have a "band gaps project" with known ID, `band_gaps_project_id`, and an associated dataset with known ID, `strehlow_cook_dataset_id`. You would like to share the "strehlow cook dataset" with another project, "semiconductors project" with a known ID `semiconductors_id`.

Reviewer: ```suggestion
Assume you have a "band gaps project" with known ID, ``band_gaps_project_id``, and an associated dataset with known ID, ``strehlow_cook_dataset_id``. You would like to share the "strehlow cook dataset" with another project, "semiconductors project" with a known ID ``semiconductors_id``.
```

Sharing a dataset:

.. code-block:: python
    
    #get the project that owns the dataset
    band_gaps_project = citrine.projects.get(band_gaps_project_id)
    
    #this shares the dataset with the ID strehlow_cook_dataset_id with the project with the ID semiconductors_id
    band_gaps_project.share(project_id=semiconductors_id, resource_type="DATASET", resource_id=strehlow_cook_dataset_id)

> Transfering a Dataset to Another Project

Reviewer: ```suggestion
Transferring a Dataset to Another Project
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> A dataset can be transfered to another project using the project from which is originates using the :func:`~citrine.resources.project.Project.transfer_resource` method on the project.

Reviewer: ```suggestion
A dataset can be transfered to another project using the :func:`~citrine.resources.project.Project.transfer_resource` method on the original project.
```

Example
^^^^^^^

> Assume you have a "band gaps project" with known ID, `band_gaps_project_id`, and an associated dataset with known ID, `strehlow_cook_dataset_id`. You would like to transfer ownership of the "strehlow cook dataset" to another project, "semiconductors project" with a known ID `semiconductors_id`.

Reviewer: ```suggestion
Assume you have a "band gaps project" with known ID, ``band_gaps_project_id``, and an associated dataset with known ID, ``strehlow_cook_dataset_id``. You would like to transfer ownership of the "strehlow cook dataset" to another project, "semiconductors project" with a known ID ``semiconductors_id``.
```
